# Изучение развития игровой индустрии

- Автор:Путилова Мария Ивановна
- Дата: 17.07.2025

### Цели и задачи проекта

Цель: изучить развитие игровой индустрии 

Задача - познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.
        1) Загрузка и знакомство с данными
        2) Проверка ошибок в данных и их предобработка
        3) Фильтрация данных
        4) Категоризация данных

### Описание данных

Данные <font color='#777778'>/datasets/new_games.csv </font> содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

Name — название игры.

Platform — название платформы.

Year of Release — год выпуска игры.

Genre — жанр игры.

NA sales — продажи в Северной Америке (в миллионах проданных копий).

EU sales — продажи в Европе (в миллионах проданных копий).

JP sales — продажи в Японии (в миллионах проданных копий).

Other sales — продажи в других странах (в миллионах проданных копий).

Critic Score — оценка критиков (от 0 до 100).

User Score — оценка пользователей (от 0 до 10).

Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.





### Содержимое проекта

 1) Загрузка и знакомство с данными
 
 2) Проверка ошибок в данных и их предобработка
 
 3) Фильтрация данных
 
 4) Категоризация данных
        

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# Импортируем библиотеку pandas

import pandas as pd


In [2]:
# Загружаем данные датсета 
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')



- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
# Применяем метод info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


- Данные содержат 11 столбцов и 16956 строк, соответствуют описанию.

- Встречаются пропуски в столбцах: Name, Year of Release, Genre, Critic Score, User Score, Rating. Наибольшее количество пропусков в столбце с оценкой критиков, наименьшее в столбце с названием игры и жанром.

- Используется 2 типа данных: float64 - 4 столбца, object - 7 столбцов. 
  В столбцах EU sales, JP sales и User Score лучше использовать тип float64. 

- Все названия столбцов отражают содержимое данных и прописаны в удобном для работе виде. 


In [4]:
количество_строк = len(df)
print(f"Количество строк: {количество_строк}")

Количество строк: 16956


---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма


- Выведем на экран названия всех столбцов датафрейма и проверим их стиль написания.


In [5]:
# Выведем на экран названия всех столбцов датафрейма и проверим их стиль написания 
print(df.columns) 

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


- Приведём все столбцы к стилю snake case. 

In [6]:
# Приводим все столбцы к стилю snake case
df.columns = df.columns.str.lower().str.replace(' ', '_')
print(df.columns) 

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


### 2.2. Типы данных



Некоректные типы данны встречаются в столбцах:  EU sales, JP sales и User Score. 
Скорее всего это связано с тем, что цифры были введены как строка. 

Для того, чтобы преобразовать данные типы к типу float64 нужно обработать из этих столбцов пропуски. 

В столбцах EU sales, JP sales пропускаов нет.


- Преобразуем пропуски в столбцах и заменим типы данных с object на float64

In [7]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce').fillna(0)
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce').fillna(0)
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce').fillna(0)

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приведём такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [8]:
df['year_of_release'] = pd.to_numeric(df['year_of_release'], errors='coerce')
df['na_sales'] = pd.to_numeric(df['na_sales'], errors='coerce')
df['other_sales'] = pd.to_numeric(df['other_sales'], errors='coerce')
df['critic_score'] = pd.to_numeric(df['critic_score'], errors='coerce')



### 2.3. Наличие пропусков в данных

- Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [9]:
# Посчитаем количество пропусков в каждом столце в абсолютных значениях 
print(df[['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']].isna().sum())

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8714
user_score            0
rating             6871
dtype: int64


In [10]:
# Посчитаем количество пропусков в каждом столце в относительных значениях 
print(df[['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']].isna().sum()/len(df))

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000000
jp_sales           0.000000
other_sales        0.000000
critic_score       0.513918
user_score         0.000000
rating             0.405225
dtype: float64


Пропуски характерны для столбцов: 'name', 'year_of_release', 'genre', 'critic_score','rating'.

Наиболшее количество пропусков в столбце 'critic_score' - 8714, это состааляет 51%. Можно предположить, что данный пропуск связан с тем, что не все игры оценивали критики. Такие пропуски лучше предобработать и заменить на какое-либо зачение. 

Также большое колмество пропусков в столбце 'ratin' - 6871, это составляет 40%. Это может быть связано с тем, что не все игры внесены врейтинг организации ESRB. Такие пропуски лучше предобработать и заменить на како-либо зачение. 

Пропуски в столбце 'year_of_release' - 275, это примерно 2%. Эии данные можно либо предоработать, либо проигнорировать, если они не играют роли в анализе. 

В столбцах 'name' и 'genre': 2, это меньше процента. Такие пропуски скорее могут быть связаны с человеческим фактором. Такие данные можно удалить или игнорировать, если они не влияют на результат исследования. 
       

In [11]:
# Удалим пропуски в столбцах: 'name', 'genre'.

df = df.dropna(subset = ['name', 'genre'])


In [12]:
# Заменим пропуски в столбце 'year_of_release' средним значением 

df['year_of_release'] = df['year_of_release'].fillna(df['year_of_release'].mean())

- Пропуски в столбце 'critic_score' заменять не стоит 

In [13]:
# Заменим пропуски в столбце 'rating' на пустую строку

df['rating'] = df['rating'].fillna('')

### 2.4. Явные и неявные дубликаты в данных


- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

Изучим уникальные значения в категориальных данных, связанных с названиями жанра игры, платформы, рейтинга и года выпуска. Проверим, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.

In [14]:
# Напишем код для вывода уникальных значений в столбцах: 'genre', 'platform', 'rating', 'year_of_release'.

print(df['genre'].unique())

print (df['platform'].unique())

print(df['rating'].unique())

print(df['year_of_release'].unique())

['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']
['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']
['E' '' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']
[2006.         1985.         2008.         2009.         1996.
 1989.         1984.         2005.         1999.         2007.
 2010.         2013.         2004.         1990.         1988.
 2002.         2001.         2011.         1998.         2015.
 2012.         2014.         1992.         1997.         1993.
 1994.         1982.         2016.         2003.         1986.
 2000.         2006.48713952 1995.         1991.         1981.
 1987.         1980.         1983.        ]


Среди жанров встречаются неявные дубликаты полностью повторяющие название с разным способом написания, например, 'Racing' и 'RACING'.

В столбце 'rating' встречается значение 'K-A', которое не относится к данному рейтингу.

В столбце 'year_of_release' так же встречается странное значение 2006.48713952, которое дублирует 2006 год, возможно это опечатка, сязанная с человеческим фактором. 

In [15]:
# Приведем жанры к нижнему регистру

df['genre'] = df['genre'].str.lower()

In [16]:
# Приведм значения столбца 'year_of_release' к типу int64

df['year_of_release'] = df['year_of_release'].astype('int64')

- Проверим наличие явных дубликатов в данных.

In [17]:
# Создаём маску для дубликатов

duplicate_mask = df.duplicated(subset=['name', 'platform', 'year_of_release', 'genre', 
                                      'na_sales', 'eu_sales', 'jp_sales', 'other_sales',
                                      'critic_score', 'user_score', 'rating'])


In [18]:
# Подсчитываем количество дубликатов

total_duplicates = duplicate_mask.sum()
print(f"Всего полных дубликатов строк: {total_duplicates}")


Всего полных дубликатов строк: 241


- Количество дубликатов = 241. 

- Удалим дубликаты

In [34]:
# Напишем код для удаления дубликатов

df = df.drop_duplicates()


- В процессе подготовки данных мы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитаем количество удалённых строк в абсолютном и относительном значениях.

In [35]:
# Посчитаем количество удаленный строк с дубликатами и пропусками в абсолютном значении

deleted_rows_abs = количество_строк-len(df)

# Посчитаем количество удаленный строк с дубликатами и пропусками в относительном значении 
deleted_rows_rel = (deleted_rows_abs / количество_строк) * 100

print(f"Удалено строк (абсолютное значение): {deleted_rows_abs}")  
print(f"Удалено строк (относительное значение): {deleted_rows_rel:.2f}%")  

Удалено строк (абсолютное значение): 243
Удалено строк (относительное значение): 1.43%


- После проведения предобработки данных напишим общий промежуточный вывод.

Привели все столбцы к стилю snake case.

Преобразовали пропуски в столбцах.

Заменили типы данных с object на float64 в столбцах: 'eu_sales', 'jp_sales', 'user_score'.
    
Посчитали количество пропусков в каждом столбце в абсолютных и относительных значениях.

Удалили пропуски в столбцах: 'name', 'genre'.
    
Заменили пропуски в столбце 'year_of_release' средним значением.

Заменили пропуски в столбце 'rating' на пустую строку.

Изучили уникальные значения в категориальных данных, связанных с названиями жанра игры, платформы, рейтинга и года выпуска. Проверили, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
Количество явных дубликатов = 241.
Удалили дубликаты.

Посчитали количество удалённых строк в абсолютном и относительном значениях.
Удалено строк (абсолютное значение): 243
Удалено строк (относительное значение): 1.43%


---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [36]:
# Создаем новый датафрейм с играми, выпущенными с 2000 по 2013 год включительно

df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)].copy()

In [22]:
# Сбрасываем индексы после фильтрации 

df_actual.reset_index(drop=True, inplace=True)

In [23]:
# Выодим результат 

print(df_actual.info())
print(df_actual.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13050 entries, 0 to 13049
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             13050 non-null  object 
 1   platform         13050 non-null  object 
 2   year_of_release  13050 non-null  int64  
 3   genre            13050 non-null  object 
 4   na_sales         13050 non-null  float64
 5   eu_sales         13050 non-null  float64
 6   jp_sales         13050 non-null  float64
 7   other_sales      13050 non-null  float64
 8   critic_score     7323 non-null   float64
 9   user_score       13050 non-null  float64
 10  rating           13050 non-null  object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.1+ MB
None
                    name platform  year_of_release     genre  na_sales  \
0             Wii Sports      Wii             2006    sports     41.36   
1         Mario Kart Wii      Wii             2008    racing     15.68   
2   

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [24]:
# Создаём пустой список для оценочных групп
evaluation_group_user = []

In [25]:
# Классификация игр по оценке пользователей с использованием циклов и условий

for score in df['user_score']:
    
    # Проверяем на NaN (пропущенные значения)
    if pd.isna(score):
        evaluation_group_user.append("нет оценки")
    
    # Проверяем диапазоны оценок (правые границы не включаются, кроме 10)
    elif 8 < score <= 10:  # 10 включается
        evaluation_group_user.append("высокая оценка")
    elif 3 <= score < 8:   # 8 не включается
        evaluation_group_user.append("средняя оценка")
    elif 0 <= score < 3:   # 3 не включается
        evaluation_group_user.append("низкая оценка")
    else:
        evaluation_group_user.append("оценка вне диапазона")
    


In [26]:
# Добавляем новый столбец в DataFrame
df['evaluation_group_user'] = evaluation_group_user

print(df['evaluation_group_user'])

0        оценка вне диапазона
1               низкая оценка
2              высокая оценка
3        оценка вне диапазона
4               низкая оценка
                 ...         
16951           низкая оценка
16952           низкая оценка
16953           низкая оценка
16954           низкая оценка
16955           низкая оценка
Name: evaluation_group_user, Length: 16713, dtype: object


- Разделим все игры по оценкам критиков и выделим такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [27]:
# Создаём пустой список для оценочных групп
evaluation_group_сritic = []

In [28]:
# Классификация игр по оценке пользователей с использованием циклов и условий

for score in df['critic_score']:
    
    # Проверяем на NaN (пропущенные значения)
    if pd.isna(score):
        evaluation_group_сritic.append("нет оценки")
    
    # Проверяем диапазоны оценок (правые границы не включаются, кроме 10)
    elif 80 < score <= 100:  # 10 включается
        evaluation_group_сritic.append("высокая оценка")
    elif 30 <= score < 80:   # 8 не включается
        evaluation_group_сritic.append("средняя оценка")
    elif 0 <= score < 30:   # 3 не включается
        evaluation_group_сritic.append("низкая оценка")
    else:
        evaluation_group_сritic.append("оценка вне диапазона")
    

In [29]:
# Добавляем новый столбец в DataFrame
df['evaluation_group_сritic'] = evaluation_group_сritic

print(df['evaluation_group_сritic'])

0              средняя оценка
1                  нет оценки
2              высокая оценка
3        оценка вне диапазона
4                  нет оценки
                 ...         
16951              нет оценки
16952              нет оценки
16953              нет оценки
16954              нет оценки
16955              нет оценки
Name: evaluation_group_сritic, Length: 16713, dtype: object


- После категоризации данных проверим результат: сгруппируем данные по выделенным категориям и посчитаем количество игр в каждой категории.

In [30]:
# Напишем код для посчета по оценке пользователей 

grouped_1 = df.groupby('evaluation_group_user').agg({'name': 'count'})

print(grouped_1)

                       name
evaluation_group_user      
высокая оценка         2298
низкая оценка          9268
оценка вне диапазона    290
средняя оценка         4857


In [31]:
# Напишем код для посчета по оценке критиков 

grouped_2 = df.groupby('evaluation_group_сritic').agg({'name': 'count'})

print(grouped_2)

                         name
evaluation_group_сritic      
высокая оценка           1775
нет оценки               8576
низкая оценка              62
оценка вне диапазона      237
средняя оценка           6063


- Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [32]:
# Сгруппируем данные по названию платформы и посчитаем количество игр на каждой платформе

df['game_count'] = df.groupby('platform').agg({'name': 'count'})

In [33]:
# Отсортируем по колисевту игр на платформе в порядке убывания и выведем 7 первых значений.

df_sorted = df.sort_values(by='game_count', ascending=False)
print(df_sorted.head(7))

                       name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii             2006        sports     41.36   
1         Super Mario Bros.      NES             1985      platform     29.08   
2            Mario Kart Wii      Wii             2008        racing     15.68   
3         Wii Sports Resort      Wii             2009        sports     15.61   
4  Pokemon Red/Pokemon Blue       GB             1996  role-playing     11.27   
5                    Tetris       GB             1989        puzzle     23.20   
6     New Super Mario Bros.       DS             2006      platform     11.28   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0     28.96      3.77         8.45          76.0         8.0      E   
1      3.58      6.81         0.77           NaN         0.0          
2     12.76      3.79         3.29          82.0         8.3      E   
3     10.93      3.28         2.95          80.0         8.0      E

---

## 5. Итоговый вывод

Напишем основной вывод и отразим, какая работы была проделана. Укажем описание среза данных и новых полей, которые добавили в исходный датасет.

Загрузили и познакомились с данными.
При знакомстве с данными было обнаружено:
- Данные содержат 11 столбцов и 16956 строк, соответствуют описанию.
- Встречаются пропуски в столбцах: Name, Year of Release, Genre, Critic Score, User Score, Rating. Наибольшее количество пропусков в столбце с оценкой критиков, наименьшее в столбце с названием игры и жанром.
- Используется 2 типа данных: float64 - 4 столбца, object - 7 столбцов. В столбцах EU sales, JP sales и User Score лучше использовать тип float64.
- Все названия столбцов отражают содержимое данных и прописаны в удобном для работе виде.


Проверили наличие ошибок в данных предобработали их.
- Вывели на экран названия всех столбцов датафрейма и проверим их стиль написания. Измени стиль написания на snake case.
- Заменили типы данных с object на float64 в столбцах: 'eu_sales', 'jp_sales', 'user_score'.
- Посчитали количество пропусков в каждом столбце в абсолютных и относительных значениях.
- Удалили пропуски в столбцах: 'name', 'genre'.
- Заменили пропуски в столбце 'year_of_release' средним значением.
- Заменили пропуски в столбце 'rating' на пустую строку.
- Посчиали и удалили явные дубликаты: Количество дубликатов = 241.

Отфильтровали данные.
- Создали новый датафрейм с играми, выпущенными с 2000 по 2013 год включительно

Выполнили категоризацию данных.
- Создали список для оценочных групп по оценке пользователей и критиков.
- Выделили топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

Добавили такие поля: 
- количество_строк (количество строк в изначальном датафрейме), 
- total_duplicates (количество дубоикатов), 
- deleted_rows_abs (количество строк, после удаления пропусков и дубликатов), 
- deleted_rows_abs (количество удаленных строк с дубликатами и пропусками в абсолютном значении), 
- deleted_rows_rel (количество удаленных строк с дубликатами и пропусками в относительном значении),
- df_actual (новый датафрейм с играми, выпущенными с 2000 по 2013 год включительно,
- evaluation_group_user (список оценочных групп пользователей),
- evaluation_group_сritic (список оценочных групп критиков),
- grouped_1 (сгруппируем данные по выделенным категориям и посчитаем количество игр в каждой категории по оценке пользователей),
- grouped_2 (сгруппируем данные по выделенным категориям и посчитаем количество игр в каждой категории по оценке критиков),
- game_count (группируем данные по названию платформы и посчитаем количество игр на каждой платформе, выводим топ 7)

